cfme/utils/events.py

```diff
@@ -335,109 +165,113 @@ def new_event(self, *attrs, **kwattrs):
+        callback = kwargs.get('callback')
+        first_event = bool(kwargs.get('first_event'))
+
+        for evt in evts:
+            if isinstance(evt, Event):
+                exp_event = {'event': evt,
+                             'callback': callback,
+                             'matched_events': [],
+                             'first_event': first_event}
+                self._events_to_listen.append(exp_event)
+                logger.info("event {} is added to listening queue.".format(evt))
+            else:
+                raise ValueError("one of events doesn't belong to Event class")
 
     def start(self):
-        logger.info('Event Listener has been started')
         self.set_last_record()
         self._stop_event.clear()
         super(EventListener, self).start()
+        logger.info('Event Listener has been started')
 
     def stop(self):
-        logger.info('Event Listener has been stopped')
         self._stop_event.set()
-
-    def run(self):
-        self.process_events()
+        logger.info('Event Listener has been stopped')
 
     @property
     def started(self):
         return super(EventListener, self).is_alive()
 
+    def run(self):
+        """ Overrides ThreadEvent run to continuously process events"""
+        self.process_events()
+
     def process_events(self):
-        """
-        processes all new db events and compares them with expected events.
-        processed events are ignored next time
+        """ Processes all new events and compares them with expected events.
+
+        Processed events are ignored next time.
         """
         while not self._stop_event.is_set():
-            events = self.get_next_portion()
-            if len(events) == 0:
-                sleep(0.2)
-                continue
-            for got_event in events:
-                logger.debug("processing event id {}".format(got_event.id))
-                got_event = Event(event_tool=self._tool).build_from_raw_event(got_event)
-                for exp_event in self._events_to_listen:
-                    if exp_event['first_event'] and len(exp_event['matched_events']) > 0:
-                        continue
+            sleep(1)
+            for exp_event in self._events_to_listen:
+
+                # Skip if event has occurred
+                if exp_event['first_event'] and len(exp_event['matched_events']):
+                    continue
 
+                matched_events = self.get_next_portion(exp_event['event'])
+
+                if not matched_events:
+                    continue
+
+                # Match events
+                got_event = None
+                for event_entity in matched_events:
+                    got_event = Event(self._appliance).build_from_entity(event_entity)
                     if exp_event['event'].matches(got_event):
                         if exp_event['callback']:
                             exp_event['callback'](exp_event=exp_event['event'], got_event=got_event)
```

Comment:
REQUIRED: there should be try/except statement since this external function and it may cause some exceptions.
If there is some exception, it will break listener.